internal/backends/sqlite/collection.go

```diff
@@ -91,6 +96,142 @@ func (c *collection) Query(ctx context.Context, params *backends.QueryParams) (*
 	}, nil
 }
 
+func prepareWhereClause(filterDoc *types.Document) (string, []any, error) {
+	if filterDoc == nil {
+		return "", []any{}, nil
+	}
+
+	iter := filterDoc.Iterator()
+	defer iter.Close()
+
+	var filters []string
+	var args []any
+
+	for {
+		k, v, err := iter.Next()
+		if errors.Is(err, iterator.ErrIteratorDone) {
+			break
+		}
+
+		if err != nil {
+			return "", nil, lazyerrors.Error(err)
+		}
+
+		queryPath := fmt.Sprintf("%s->'$.\"%s\"'", metadata.DefaultColumn, k)
+
+		if k == "_id" {
+			queryPath = metadata.IDColumn
+		}
+
+		// don't pushdown $comment, it's attached to query in handlers
+		if strings.HasPrefix(k, "$") {
+			continue
+		}
+
+		path, err := types.NewPathFromString(k)
+
+		var pe *types.PathError
+
+		switch {
+		case err == nil:
+			// Handle dot notation.
+			// TODO https://github.com/FerretDB/FerretDB/issues/2069
+			if path.Len() > 1 {
+				continue
+			}
+		case errors.As(err, &pe):
+			// ignore empty key error, otherwise return error
+			if pe.Code() != types.ErrPathElementEmpty {
+				return "", nil, lazyerrors.Error(err)
+			}
+		default:
+			panic("Invalid error type: PathError expected")
+		}
+
+		switch v := v.(type) {
+		case *types.Array, *types.Document, types.Binary, types.NullType, types.Regex, types.Timestamp:
+			// type not supported for pushdown
+			continue
+
+		case int64:
+			maxSafeDouble := int64(types.MaxSafeDouble)
+
+			var sql string
+
+			// If value cannot be safe double, fetch all numbers out of the safe range.
+			switch {
+			case v > maxSafeDouble:
+				sql = ` > ?`
+				v = maxSafeDouble
+
+			case v < -maxSafeDouble:
+				sql = `< ?`
+				v = -maxSafeDouble
+			default:
+				// don't change the default eq query
+			}
+
+			subquery := fmt.Sprintf(`EXISTS (SELECT value FROM json_each(%v) WHERE value %s)`, queryPath, sql) // TODO sanitize the key
+
+			filters = append(filters, subquery)
+			args = append(args, v)
+
+		case float64:
+			var sql string
+			switch {
+			case v > types.MaxSafeDouble:
+				sql = ` > ?`
+				v = types.MaxSafeDouble
+
+			case v < -types.MaxSafeDouble:
+				sql = ` < ?`
+				v = -types.MaxSafeDouble
+			default:
+				// don't change the default eq query
+			}
+
+			subquery := fmt.Sprintf(`EXISTS (SELECT value FROM json_each(%v) WHERE value %s)`, queryPath, sql) // TODO sanitize the key
+
+			filters = append(filters, subquery)
+			args = append(args, v)
+
+		case time.Time:
+			subquery := fmt.Sprintf(`EXISTS (SELECT value FROM json_each(%v) WHERE value = ?)`, queryPath) // TODO sanitize the key
+
+			filters = append(filters, subquery)
+			val := v.UnixMilli()
+
+			args = append(args, val)
```

Comment:
@AlekSi here we need to convert time.Time to int, by using UnixMilli. I'm not quiet sure if this parsing logic should be here, and using sjson package for that seems more relevant, though the Marshal functions return strings.

Reply:
But why `MarshalSingleValue` can't be used there?

Reply:
It returns slice of bytes, which can be parsed to string, and SQLite library will search for strings and not numbers

Reply:
```
test> db.test.find()
[
  {
    _id: ObjectId("65097913f4928d7f17fa78b6"),
    t: ISODate("2023-09-19T10:33:55.233Z")
  },
  {
    _id: ObjectId("65097915f4928d7f17fa78b7"),
    t: ISODate("2023-09-19T10:33:57.941Z")
  },
  {
    _id: ObjectId("65097918f4928d7f17fa78b8"),
    t: ISODate("2023-09-19T10:34:00.403Z")
  },
  {
    _id: ObjectId("65097a40e627ae53f9f9604d"),
    t: [ ISODate("2023-09-19T10:38:56.881Z") ]
  },
  {
    _id: ObjectId("65097a41e627ae53f9f9604e"),
    t: [ ISODate("2023-09-19T10:38:57.787Z") ]
  },
  {
    _id: ObjectId("65097a43e627ae53f9f9604f"),
    t: [ ISODate("2023-09-19T10:38:59.310Z") ]
  }
]
```

```
sqlite> SELECT * FROM test_afd071e5 WHERE EXISTS (SELECT value FROM json_each(_ferretdb_sjson->'$.t') WHERE value = 1695119635233);
_ferretdb_sjson
{"$s":{"p":{"_id":{"t":"objectId"},"t":{"t":"date"}},"$k":["_id","t"]},"_id":"65097913f4928d7f17fa78b6","t":1695119635233}
sqlite> SELECT * FROM test_afd071e5 WHERE EXISTS (SELECT value FROM json_each(_ferretdb_sjson->'$.t') WHERE value = 1695119939310);
_ferretdb_sjson
{"$s":{"p":{"_id":{"t":"objectId"},"t":{"t":"array","i":[{"t":"date"}]}},"$k":["_id","t"]},"_id":"65097a43e627ae53f9f9604f","t":[1695119939310]}
```

So that seems to work, and `MarshalSingleValue` would return a JSON representation of that time as a JSON slice of bytes / string without quotes, so that should work, no?

Reply:
Then, let's do `value = %s` without a placeholder. It still sounds sus, TBH.

Reply:
That's what is produced when we provide a string value into the SQLite
```
2023-09-19T12:54:20.219+0200    DEBUG   sqlite.test     fsql/db.go:77   >>> SELECT _ferretdb_sjson FROM "values_34474c3b" WHERE EXISTS (SELECT value FROM json_each(_ferretdb_sjson->'$."v"') WHERE value = ?)       {"args": ["1695119939310"]}
```

Reply:
> Then, let's do value = %s without a placeholder. It still sounds sus, TBH.

What about sanitizing?

Reply:
I think we should use `json`, not `value`, as we use `->` operator for indexes that returns JSON representation as text:

```
sqlite> SELECT _ferretdb_sjson->'$.t', typeof(_ferretdb_sjson->'$.t') FROM test_afd071e5;
┌────────────────────────┬────────────────────────────────┐
│ _ferretdb_sjson->'$.t' │ typeof(_ferretdb_sjson->'$.t') │
├────────────────────────┼────────────────────────────────┤
│ 1695119635233          │ text                           │
│ 1695119637941          │ text                           │
│ 1695119640403          │ text                           │
│ [1695119936881]        │ text                           │
│ [1695119937787]        │ text                           │
│ [1695119939310]        │ text                           │
└────────────────────────┴────────────────────────────────┘
sqlite> SELECT value, typeof(value) FROM test_afd071e5, json_each(_ferretdb_sjson->'$.t');
┌───────────────┬───────────────┐
│     value     │ typeof(value) │
├───────────────┼───────────────┤
│ 1695119635233 │ integer       │
│ 1695119637941 │ integer       │
│ 1695119640403 │ integer       │
│ 1695119936881 │ integer       │
│ 1695119937787 │ integer       │
│ 1695119939310 │ integer       │
└───────────────┴───────────────┘
sqlite> SELECT json, typeof(json) FROM test_afd071e5, json_each(_ferretdb_sjson->'$.t');
┌─────────────────┬──────────────┐
│      json       │ typeof(json) │
├─────────────────┼──────────────┤
│ 1695119635233   │ text         │
│ 1695119637941   │ text         │
│ 1695119640403   │ text         │
│ [1695119936881] │ text         │
│ [1695119937787] │ text         │
│ [1695119939310] │ text         │
└─────────────────┴──────────────┘
```

Reply:
Or maybe not:

```
sqlite> SELECT json, typeof(json) FROM test_afd071e5, json_each(_ferretdb_sjson->'$.t');
┌───────────────────────────────┬──────────────┐
│             json              │ typeof(json) │
├───────────────────────────────┼──────────────┤
│ 1695119635233                 │ text         │
│ 1695119637941                 │ text         │
│ 1695119640403                 │ text         │
│ [1695119936881]               │ text         │
│ [1695119937787]               │ text         │
│ [1695119939310]               │ text         │
│ [1695121115855,1695121115855] │ text         │
│ [1695121115855,1695121115855] │ text         │
└───────────────────────────────┴──────────────┘
```

I think I will leave you with that now :) Just be careful with type-checking.

---

> What about sanitizing?

There is no need to sanitize a number. We should not do that in all cases

Reply:
Unfortunately it's not that easy to use json, as it treats arrays as whole:
```
sqlite> SELECT json, value FROM "values_34474c3b" join json_each(_ferretdb_sjson->'$."v"');
json|value
1|1
[1,2,1]|1
[1,2,1]|2
[1,2,1]|1
2|2
"foo"|foo
"65046bcc4ed83bef4404a73a"|65046bcc4ed83bef4404a73a
"65046bcc4ed83bef4404a73a"|65046bcc4ed83bef4404a73a
"65046bcc4ed83bef4404a73a"|65046bcc4ed83bef4404a73a
1695119939310|1695119939310
true|1
["65046bcc4ed83bef4404a73a"]|65046bcc4ed83bef4404a73a
```

Because of that we cannot query the items of array